core/genesis.go

```diff
@@ -403,19 +403,18 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, triedb *triedb.Database, g
 		return chainCfg, block.Hash(), nil, nil
 	}
 
-	// TODO: Reinstate transitioned network check. Not sure where exactly to put it after this refactor.
-	// If the bedrock block is not 0, that implies that the network was migrated at the bedrock block.
-	// In this case the genesis state may not be in the state database (e.g. op-geth is performing a snap
-	// sync without an existing datadir) & even if it were, would not be useful as op-geth is not able to
-	// execute the pre-bedrock STF.
-	// transitionedNetwork := genesis != nil && genesis.Config != nil && genesis.Config.BedrockBlock != nil && genesis.Config.BedrockBlock.Uint64() != 0
+	// OP-Stack note: if it's a bedrock-transition genesis,
+	// then only commit to DB if it has a StateHash to work with,
+	// as the genesis config accounts map will not be usable.
+	transitionedNetwork := genesis != nil && genesis.Config != nil && genesis.Config.BedrockBlock != nil && genesis.Config.BedrockBlock.Uint64() != 0
+	canCommitGenesis := !transitionedNetwork || genesis.StateHash != nil
 
 	// Commit the genesis if the genesis block exists in the ancient database
 	// but the key-value database is empty without initializing the genesis
 	// fields. This scenario can occur when the node is created from scratch
 	// with an existing ancient store.
 	storedCfg := rawdb.ReadChainConfig(db, ghash)
-	if storedCfg == nil {
+	if storedCfg == nil && canCommitGenesis {
```

Comment:
Is there an easy way for us to add a test for the special behavior of Bedrock-transitioned networks?

Reply:
No clean way to test the legacy bedrock datadir without introducing a huge data dependency I am afraid of. But I'll download the datadir and test it.